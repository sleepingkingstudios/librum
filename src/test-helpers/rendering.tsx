import * as React from 'react';
import { Provider as ReduxProvider } from 'react-redux';
import {
  MemoryRouter,
  Routes,
  Route
} from "react-router-dom";
import {
  RenderOptions,
  RenderResult,
  render,
} from '@testing-library/react';

import {
  createStore,
  storeType,
} from './store';
import {
  ThemeContext,
  defaultTheme,
} from '@themes/context';
import type { Theme } from '@themes';

interface customRenderOptions extends RenderOptions {
  initialEntries?: Array<string>;
  router?: true | withRouterOptions;
  store?: (storeType | true);
  theme?: (Theme | true);
}

interface withRouterOptions {
  initialEntries?: Array<string>;
  path?: string;
}

export const withRouter = (
  component: React.ReactElement,
  {
    initialEntries = ['/'],
    path,
  }: withRouterOptions,
): React.ReactElement => {
  let wrapped = component;

  if (path) {
    wrapped = (
      <Routes>
        <Route path={path} element={wrapped} />
      </Routes>
    );
  }

  return (
    <MemoryRouter initialEntries={initialEntries}>
      { wrapped }
    </MemoryRouter>
  );
};

export const withStore = (
  component: React.ReactElement,
  { store }: { store: (storeType | true) }
): React.ReactElement => {
  let configuredStore: storeType;

  if (store === true) {
    configuredStore = createStore().store;
  } else {
    configuredStore = store;
  }

  return (
    <ReduxProvider store={configuredStore}>
      { component }
    </ReduxProvider>
  );
};

export const withTheme = (
  component: React.ReactElement,
  { theme }: { theme: (Theme | true) }
): React.ReactElement => {
  let configuredTheme: Theme;

  if (theme === true) {
    configuredTheme = defaultTheme;
  } else {
    configuredTheme = theme;
  }

  return (
    <ThemeContext.Provider value={configuredTheme}>
      { component }
    </ThemeContext.Provider>
  );
};

const customRender = (
  component: React.ReactElement,
  options?: customRenderOptions,
  strictMode = true,
): RenderResult => {
  let wrapped: React.ReactElement = component;

  if (strictMode) {
    wrapped = (<React.StrictMode>{ wrapped }</React.StrictMode>);
  }

  if (!options) { return render(wrapped); }

  if ('router' in options && options.router) {
    if (options.router === true) {
      wrapped = withRouter(wrapped, { initialEntries: options.initialEntries });
    } else {
      wrapped = withRouter(
        wrapped,
        {
          initialEntries: options.initialEntries,
          ...options.router,
        },
      );
    }
  }

  if ('store' in options) {
    wrapped = withStore(wrapped, { store: options.store });
  }

  if ('theme' in options) {
    wrapped = withTheme(wrapped, { theme: options.theme });
  }

  return render(wrapped, options);
};

export { customRender as render };
