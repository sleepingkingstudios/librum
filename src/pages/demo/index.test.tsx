import * as React from 'react';

import { DemoPage } from './index';
import { render } from '@test-helpers/rendering';

// eslint-disable-next-line @typescript-eslint/no-unsafe-return
jest.mock('@components/page', () => require('@components/page/mocks'));
// eslint-disable-next-line @typescript-eslint/no-unsafe-return
jest.mock('@core/navigation', () => require('@core/navigation/mocks'));

describe('<DemoPage />', () => {
  it('should match the snapshot', () => {
    const { asFragment } = render(<DemoPage />);

    expect(asFragment()).toMatchSnapshot();
  });
});
