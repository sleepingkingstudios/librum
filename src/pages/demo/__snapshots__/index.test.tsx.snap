// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DemoPage /> should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="mock-page"
  >
    <ul
      aria-label="navigation"
    >
      <li
        aria-label="navigation-item"
      >
        Nav Item @ /nav-item
      </li>
      <li>
        Dropdown
        <ul>
          <li
            aria-label="navigation-item"
          >
            Dropdown Item @ /dropdown-item
          </li>
        </ul>
      </li>
    </ul>
    <h1>
      Demo
    </h1>
    <p>
      This page serves as a demonstration for the styled components defined for Librum.
    </p>
    <hr
      class="hr-muted m-5"
    />
    <h2>
      Headings
    </h2>
    <h3>
      Heading 3
    </h3>
    <h4>
      Heading 4
    </h4>
    <h5>
      Heading 5
    </h5>
    <h6>
      Heading 6
    </h6>
    <hr
      class="hr-muted m-5"
    />
    <h2>
      Headings With Text
    </h2>
    <p>
      This is separator text.
    </p>
    <h3>
      Heading 3
    </h3>
    <p>
      This is separator text.
    </p>
    <h4>
      Heading 4
    </h4>
    <p>
      This is separator text.
    </p>
    <h5>
      Heading 5
    </h5>
    <p>
      This is separator text.
    </p>
    <h6>
      Heading 6
    </h6>
    <p>
      This is separator text.
    </p>
    <h3>
      <span
        class="text-danger mr-1"
      >
        Headings
      </span>
      <span
        class="text-info mx-1"
      >
        Can
      </span>
      <span
        class="text-muted mx-1"
      >
        Also
      </span>
      <span
        class="text-primary mx-1"
      >
        Have
      </span>
      <span
        class="text-success mx-1"
      >
        Colorful
      </span>
      <span
        class="text-warning mx-1"
      >
        Text
      </span>
    </h3>
    <hr
      class="hr-muted m-5"
    />
    <h3>
      Typography
    </h3>
    <p>
      This is a short paragraph.
    </p>
    <p>
      This will be a long paragraph.
    </p>
    <p>
      This will be a long paragraph.
    </p>
    <h4>
      Text Colors
    </h4>
    <p>
      This paragraph contains plain text.
    </p>
    <p>
      This paragraph contains 
      <span
        class="text-danger"
      >
        danger text
      </span>
      .
    </p>
    <p>
      This paragraph contains 
      <span
        class="text-info"
      >
        info text
      </span>
      .
    </p>
    <p>
      This paragraph contains 
      <span
        class="text-muted"
      >
        muted text
      </span>
      .
    </p>
    <p>
      This paragraph contains 
      <span
        class="text-primary"
      >
        primary text
      </span>
      .
    </p>
    <p>
      This paragraph contains 
      <span
        class="text-success"
      >
        success text
      </span>
      .
    </p>
    <p>
      This paragraph contains 
      <span
        class="text-warning"
      >
        warning text
      </span>
      .
    </p>
    <h4>
      Links
    </h4>
    <p>
      This is a 
      <a>
        basic link
      </a>
      .
    </p>
    <p>
      This is a 
      <a
        class="link-danger"
      >
        danger link
      </a>
      .
    </p>
    <p>
      This is a 
      <a
        class="link-muted"
      >
        muted link
      </a>
      .
    </p>
    <hr
      class="hr-muted m-5"
    />
    <h2>
      Data
    </h2>
    <p>
      This is a data list.
    </p>
    <dl
      class="data-list"
    >
      <div
        class="data-list-item"
      >
        <dt
          class="data-list-label"
        >
          Name
        </dt>
        <dd
          class="data-list-value"
        >
          Magic Missile
        </dd>
      </div>
      <div
        class="data-list-item"
      >
        <dt
          class="data-list-label"
        >
          Description
        </dt>
        <dd
          class="data-list-value"
        >
          <em>
            (unknown)
          </em>
        </dd>
      </div>
    </dl>
    <hr
      class="hr-muted m-5"
    />
    <h2>
      Forms
    </h2>
    <p>
      These are sample buttons:
    </p>
    <p>
      <button
        class="btn"
        type="button"
      >
        Default
      </button>
      <button
        class="btn btn-primary"
        type="button"
      >
        Primary
      </button>
      <button
        class="btn btn-info"
        type="button"
      >
        Info
      </button>
      <button
        class="btn btn-success"
        type="button"
      >
        Success
      </button>
      <button
        class="btn btn-warning"
        type="button"
      >
        Warning
      </button>
      <button
        class="btn btn-danger"
        type="button"
      >
        Danger
      </button>
      <button
        class="btn btn-muted"
        type="button"
      >
        Muted
      </button>
    </p>
    <p>
      These are outline buttons:
    </p>
    <p>
      <button
        class="btn btn-outline"
        type="button"
      >
        Default
      </button>
      <button
        class="btn btn-outline btn-primary"
        type="button"
      >
        Primary
      </button>
      <button
        class="btn btn-outline btn-info"
        type="button"
      >
        Info
      </button>
      <button
        class="btn btn-outline btn-success"
        type="button"
      >
        Success
      </button>
      <button
        class="btn btn-outline btn-warning"
        type="button"
      >
        Warning
      </button>
      <button
        class="btn btn-outline btn-danger"
        type="button"
      >
        Danger
      </button>
      <button
        class="btn btn-outline btn-muted"
        type="button"
      >
        Muted
      </button>
    </p>
    <p>
      Buttons come in a range of sizes:
    </p>
    <p>
      <button
        class="btn btn-outline btn-lg"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-truck-monster pr-2"
          data-icon="truck-monster"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 640 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M288 64v64H416L368 64H288zM419.2 25.6L496 128h80c17.7 0 32 14.3 32 32v64c17.7 0 32 14.3 32 32s-14.3 32-32 32c-29.2-38.9-75.7-64-128-64s-98.8 25.1-128 64H288c-29.2-38.9-75.7-64-128-64s-98.8 25.1-128 64c-17.7 0-32-14.3-32-32s14.3-32 32-32V160c0-17.7 14.3-32 32-32H224V48c0-26.5 21.5-48 48-48h96c20.1 0 39.1 9.5 51.2 25.6zM152 256h16c12.1 0 22.1 8.9 23.8 20.6c7.6 2.2 14.9 5.3 21.7 9c9.4-7 22.8-6.3 31.3 2.3l11.3 11.3c8.6 8.6 9.3 21.9 2.3 31.3c3.7 6.8 6.8 14.1 9 21.7c11.6 1.7 20.6 11.7 20.6 23.8v16c0 12.1-8.9 22.1-20.6 23.8c-2.2 7.6-5.3 14.9-9 21.7c7 9.4 6.3 22.8-2.3 31.3l-11.3 11.3c-8.6 8.6-21.9 9.3-31.3 2.2c-6.8 3.7-14.1 6.8-21.7 9C190.1 503.1 180.1 512 168 512H152c-12.1 0-22.1-8.9-23.8-20.6c-7.6-2.2-14.9-5.3-21.7-9c-9.4 7.1-22.8 6.3-31.3-2.2L63.8 468.9c-8.6-8.6-9.3-21.9-2.3-31.3c-3.7-6.9-6.8-14.1-9-21.8C40.9 414.1 32 404.1 32 392V376c0-12.1 8.9-22.1 20.6-23.8c2.2-7.6 5.3-14.9 9-21.8c-7-9.4-6.3-22.8 2.3-31.3l11.3-11.3c8.6-8.6 21.9-9.3 31.3-2.3c6.8-3.7 14.1-6.8 21.7-9c1.7-11.6 11.7-20.6 23.8-20.6zm8 176a48 48 0 1 0 0-96 48 48 0 1 0 0 96zM448.2 276.6c1.7-11.6 11.7-20.6 23.8-20.6h16c12.1 0 22.1 8.9 23.8 20.6c7.6 2.2 14.9 5.3 21.8 9c9.4-7 22.8-6.3 31.3 2.3l11.3 11.3c8.6 8.6 9.3 21.9 2.2 31.3c3.7 6.8 6.8 14.1 9 21.7c11.6 1.7 20.6 11.7 20.6 23.8v16c0 12.1-8.9 22.1-20.6 23.8c-2.2 7.6-5.3 14.9-9 21.7c7 9.4 6.3 22.8-2.2 31.3l-11.3 11.3c-8.6 8.6-21.9 9.3-31.3 2.2c-6.9 3.7-14.1 6.8-21.8 9C510.1 503.1 500.1 512 488 512H472c-12.1 0-22.1-8.9-23.8-20.6c-7.6-2.2-14.9-5.3-21.7-9c-9.4 7.1-22.8 6.3-31.3-2.2l-11.3-11.3c-8.6-8.6-9.3-21.9-2.2-31.3c-3.7-6.9-6.8-14.1-9-21.8C360.9 414.1 352 404.1 352 392V376c0-12.1 8.9-22.1 20.6-23.8c2.2-7.6 5.3-14.9 9-21.8c-7-9.4-6.3-22.8 2.2-31.3l11.3-11.3c8.6-8.6 21.9-9.3 31.3-2.3c6.8-3.7 14.1-6.8 21.7-9zM528 384a48 48 0 1 0 -96 0 48 48 0 1 0 96 0z"
            fill="currentColor"
          />
        </svg>
        Large
      </button>
      <button
        class="btn btn-outline btn-md"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-bicycle pr-2"
          data-icon="bicycle"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 640 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M312 32c-13.3 0-24 10.7-24 24s10.7 24 24 24h25.7l34.6 64H222.9l-27.4-38C191 99.7 183.7 96 176 96H120c-13.3 0-24 10.7-24 24s10.7 24 24 24h43.7l22.1 30.7-26.6 53.1c-10-2.5-20.5-3.8-31.2-3.8C57.3 224 0 281.3 0 352s57.3 128 128 128c65.3 0 119.1-48.9 127-112h49c8.5 0 16.3-4.5 20.7-11.8l84.8-143.5 21.7 40.1C402.4 276.3 384 312 384 352c0 70.7 57.3 128 128 128s128-57.3 128-128s-57.3-128-128-128c-13.5 0-26.5 2.1-38.7 6L375.4 48.8C369.8 38.4 359 32 347.2 32H312zM458.6 303.7l32.3 59.7c6.3 11.7 20.9 16 32.5 9.7s16-20.9 9.7-32.5l-32.3-59.7c3.6-.6 7.4-.9 11.2-.9c39.8 0 72 32.2 72 72s-32.2 72-72 72s-72-32.2-72-72c0-18.6 7-35.5 18.6-48.3zM133.2 368h65c-7.3 32.1-36 56-70.2 56c-39.8 0-72-32.2-72-72s32.2-72 72-72c1.7 0 3.4 .1 5.1 .2l-24.2 48.5c-9 18.1 4.1 39.4 24.3 39.4zm33.7-48l50.7-101.3 72.9 101.2-.1 .1H166.8zm90.6-128H365.9L317 274.8 257.4 192z"
            fill="currentColor"
          />
        </svg>
        Medium
      </button>
      <button
        class="btn btn-outline btn-sm"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-paper-plane pr-2"
          data-icon="paper-plane"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M498.1 5.6c10.1 7 15.4 19.1 13.5 31.2l-64 416c-1.5 9.7-7.4 18.2-16 23s-18.9 5.4-28 1.6L284 427.7l-68.5 74.1c-8.9 9.7-22.9 12.9-35.2 8.1S160 493.2 160 480V396.4c0-4 1.5-7.8 4.2-10.7L331.8 202.8c5.8-6.3 5.6-16-.4-22s-15.7-6.4-22-.7L106 360.8 17.7 316.6C7.1 311.3 .3 300.7 0 288.9s5.9-22.8 16.1-28.7l448-256c10.7-6.1 23.9-5.5 34 1.4z"
            fill="currentColor"
          />
        </svg>
        Small
      </button>
    </p>
    <div
      class="flex flex-col sm:flex-row"
    >
      <h3>
        Heading With Buttons
      </h3>
      <ul
        class="justify-end flex flex-col grow sm:flex-row btn-list"
      >
        <button
          class="btn btn-info"
          type="button"
        >
          Show
        </button>
        <button
          class="btn btn-outline btn-primary"
          type="button"
        >
          Update
        </button>
        <button
          class="btn btn-outline btn-danger"
          type="button"
        >
          Destroy
        </button>
      </ul>
    </div>
    <p>
      Headings can have attached buttons as well.
    </p>
    <form
      action="#"
      class="relative"
    >
      <div
        class="form-field"
      >
        <label
          class="form-label"
          for="version"
        >
          Version
        </label>
        <input
          class="form-input"
          id="version"
          name="version"
          type="text"
          value="red"
        />
      </div>
      <div
        class="form-field"
      >
        <label
          class="form-label"
          for="generation"
        >
          Generation
        </label>
        <input
          class="form-input"
          id="generation"
          name="generation"
          type="text"
          value="1"
        />
      </div>
      <button
        class="form-button"
        type="submit"
      >
        Submit
      </button>
    </form>
    <br />
    <p>
      This is a grid form.
    </p>
    <form
      action="#"
      class="relative"
    >
      <div
        class="gap-3 sm:grid sm:grid-cols-3 lg:grid-cols-6"
      >
        <div
          class="form-field"
        >
          <label
            class="form-label"
            for="strength"
          >
            Strength
          </label>
          <input
            class="form-input"
            id="strength"
            name="strength"
            type="text"
            value=""
          />
        </div>
        <div
          class="form-field"
        >
          <label
            class="form-label"
            for="dexterity"
          >
            Dexterity
          </label>
          <input
            class="form-input"
            id="dexterity"
            name="dexterity"
            type="text"
            value=""
          />
        </div>
        <div
          class="form-field"
        >
          <label
            class="form-label"
            for="constitution"
          >
            Constitution
          </label>
          <input
            class="form-input"
            id="constitution"
            name="constitution"
            type="text"
            value=""
          />
        </div>
        <div
          class="form-field"
        >
          <label
            class="form-label"
            for="intelligence"
          >
            Intelligence
          </label>
          <input
            class="form-input"
            id="intelligence"
            name="intelligence"
            type="text"
            value=""
          />
        </div>
        <div
          class="form-field"
        >
          <label
            class="form-label"
            for="wisdom"
          >
            Wisdom
          </label>
          <input
            class="form-input"
            id="wisdom"
            name="wisdom"
            type="text"
            value=""
          />
        </div>
        <div
          class="form-field"
        >
          <label
            class="form-label"
            for="charisma"
          >
            Charisma
          </label>
          <input
            class="form-input"
            id="charisma"
            name="charisma"
            type="text"
            value=""
          />
        </div>
        <div
          class="form-field col-span-2"
        >
          <label
            class="form-label"
            for="armor_class"
          >
            Armor Class
          </label>
          <input
            class="form-input"
            id="armor_class"
            name="armor_class"
            type="text"
            value=""
          />
        </div>
        <div
          class="form-field col-span-2"
        >
          <label
            class="form-label"
            for="saving_throws"
          >
            Saving Throws
          </label>
          <input
            class="form-input"
            id="saving_throws"
            name="saving_throws"
            type="text"
            value=""
          />
        </div>
        <div
          class="form-field col-span-2"
        >
          <label
            class="form-label"
            for="skills"
          >
            Skills
          </label>
          <input
            class="form-input"
            id="skills"
            name="skills"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="gap-3 sm:grid sm:grid-cols-2 lg:grid-cols-4"
        style="direction: rtl;"
      >
        <button
          class="form-button col-span-2"
          type="submit"
        >
          Submit
        </button>
        <button
          class="form-button"
          type="button"
        >
          Cancel
        </button>
      </div>
    </form>
    <ul
      aria-label="breadcrumbs"
    >
      <li
        aria-label="breadcrumb"
      >
        Home @ /
      </li>
      <li
        aria-label="breadcrumb"
      >
        Admin
      </li>
      <li
        aria-label="breadcrumb"
      >
        Demo @ /demo
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
